README.md
# Simple HTTP Server

[![Test Go](https://github.com/Cyb3r-Jak3/simple_http_server/actions/workflows/golang.yml/badge.svg)](https://github.com/Cyb3r-Jak3/simple_http_server/actions/workflows/golang.yml) [![Publish Docker](https://github.com/Cyb3r-Jak3/simple_http_server/actions/workflows/docker.yml/badge.svg)](https://github.com/Cyb3r-Jak3/simple_http_server/actions/workflows/docker.yml) [![DeepSource](https://deepsource.io/gh/Cyb3r-Jak3/simple_http_server.svg/?label=active+issues&show_trend=true)](https://deepsource.io/gh/Cyb3r-Jak3/simple_http_server/?ref=repository-badge)[![Go Report Card](https://goreportcard.com/badge/github.com/Cyb3r-Jak3/simple_http_server)](https://goreportcard.com/report/github.com/Cyb3r-Jak3/simple_http_server)

This is a simple HTTP server which I create while learning about HTTP requests in GoLang. Mocked data is generated using [gofakeit](https://github.com/brianvoe/gofakeit/v6)

Releases are signed with my [release key](https://gist.github.com/Cyb3r-Jak3/8a9ba09406d991d5bab0d677b1af799d)

## Running

To run this program you can download either the docker image or a [release binary](https://github.com/Cyb3r-Jak3/simple_http_server/releases/latest).

### Customizing

#### Host

By default the host is `0.0.0.0` and it can be changed with an environment variable of `HOST`

#### Port

By default the port is `8090` and it can be changed with an environment variable of `PORT`

### Docker

There are images on both Github and Dockerhub.

**Docker:**
`docker run -d -p 8090:8090 cyb3rjak3/simple_http_server`

**Github:**
`docker run -d -p 8090:8090 ghcr.io/cyb3r-jak3/simple_http_server`

## Routes

The routes that are available are:

### /

Returns a simple hello message

### /headers

Echos back the headers that were sent with the request

### /post/json

Echos back the JSON that was sent with the request

### /post/file/form

Saves a posted multipart/form-data file to the server.

**Required**: The key of the posted file to be "file"

### /post/file/{name}

Saves a post binary file with the name

**Required**:

  {name}: Query string set that will set the name of the file

### /get/json/{rows}

Returns a number of JSON rows

**Optional**:

    {rows}: Return the number of rows to generate. Default is 10

### /get/xml/{rows}

Returns a number of XML rows

**Optional**:

    {rows}: Return the number of rows to generate. Default is 10

### /get/ipv4

Returns an IPv4 address

### /get/ipv6

Returns an IPv6 address

### /get/base64

Returns a paragraph that has been base64 encoded

### /get/csv/{rows}

Returns a csv file

**Optional**:

    {rows}: Return the number of rows to generate. Default is 10

### /get/image{type}/{height}/{width}

Returns an image either JPG, PNG or the url to download

**Optional**:

    - {type}: Type to generate. Valid options are url, png, jpg
    - {height}: Height in pixels of the image
    - {width}: Width in pixels of the image. If none is given a square image of height will be generated

### /get/uuid

Returns a UUID4 as a string

### /cookies/get

Returns the cookies of the request

### /cookies/set/{name}/{value}

Set a cookie in the response

**Required**:

    - {name}: Name of the Cookie
    - {value}: Value of the cookie

### /cookies/clear

Clears all the cookies from a request

### "/status/{code}", StatusCode)

Return a request with status code

**Optional**:
    - {code}: HTTP status code to return

### "/redirect/{code}"

Return to the constant domain

**Optional**:

    - {code}: HTTP redirect code between 300 and 307

### "/auth/basic/{username}/{password}"

Generates HTTP Basic auth

**Required**:

    - {username}: User name to login with
    - {password}: Password to login with

### "/auth/basic/bad"

HTTP basic auth with username and password of admin


CHANGELOG.md

# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.3.1 - 2021-07-06

### Added

- Merge of basic auth #15

### Changed

- Fixed docker manifest issue #9

## 1.3.0 - 2021-05-18

### Added

- Added use of `HOST` and `PORT` environment variables

## 1.1.1 - 2021-04-20

### Added

- Release to docker hub
- GoReleaser Docker Manifest
- Security code scanning
- gosec scanning
- GetCSV function

### Changed

- Matching docker container tags
- Use crypto/rand for secure number generation

## 1.1.0 - 2021-04-17

### Added

- GetXML, GetIPv4, GetIPv6, and GetBase64 functions
- GoReleaser PR Check

### Changed

- Releasing docker images with GoReleaser

## 1.0.0 - 2021-04-15

### Added

- All code related to routes
- Go Linting
- Docker image building
- GoLang release building


get_test.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"net/http"
	"regexp"
	"testing"

	"github.com/brianvoe/gofakeit/v6"
)

func TestGETJSON(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/json", nil)
	rr := executeRequest(r, GetJSON)
	var jsonObjs interface{}
	err := json.Unmarshal(rr.Body.Bytes(), &jsonObjs)
	if err != nil {
		t.Errorf("Unable to marshal JSON. %s\n", err)
	}
	objSlice, _ := jsonObjs.([]interface{})
	if len(objSlice) != 10 {
		t.Errorf("Should be 10 JSON objects. Got %d\n", len(objSlice))
	}
	checkResponse(t, rr, http.StatusOK)
	r, _ = http.NewRequest("GET", "/get/json/5", nil)
	rr = executeVarsRequest("/get/json/{rows}", r, GetJSON)
	checkResponse(t, rr, http.StatusOK)
	r, _ = http.NewRequest("GET", "/get/json/hello", nil)
	rr = executeVarsRequest("/get/json/{rows}", r, GetJSON)
	checkResponse(t, rr, http.StatusBadRequest)
}

func TestGetImagePNG(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/image/png", nil)
	rr := executeVarsRequest("/get/image/{type}", r, GetImage)
	checkResponse(t, rr, http.StatusOK)
	r, _ = http.NewRequest("GET", "/get/image", nil)
	rr = executeRequest(r, GetImage)
	checkResponse(t, rr, http.StatusOK)
	r, _ = http.NewRequest("GET", "/get/image/png/200", nil)
	rr = executeVarsRequest("/get/image/{type}/{height}", r, GetImage)
	checkResponse(t, rr, http.StatusOK)
	r, _ = http.NewRequest("GET", "/get/image/png/200/200", nil)
	rr = executeVarsRequest("/get/image/{type}/{height}/{width}", r, GetImage)
	checkResponse(t, rr, http.StatusOK)
	r, _ = http.NewRequest("GET", "/get/image/200/200", nil)
	rr = executeVarsRequest("/get/image/{type}/{height}", r, GetImage)
	checkResponse(t, rr, http.StatusOK)
}

func TestGetImageJPG(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/image/jpg", nil)
	rr := executeVarsRequest("/get/image/{type}", r, GetImage)
	checkResponse(t, rr, http.StatusOK)

}
func TestGetImageURL(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/image/url", nil)
	rr := executeVarsRequest("/get/image/{type}", r, GetImage)
	checkResponse(t, rr, http.StatusOK)
}

func TestGetImageFake(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/image/fake", nil)
	rr := executeVarsRequest("/get/image/{type}", r, GetImage)
	checkResponse(t, rr, http.StatusOK)
}

func TestGetUUID(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/uuid", nil)
	rr := executeRequest(r, GetUUID)
	// Using regex because I don't want to import a uuid package for testing
	reg := regexp.MustCompile("^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[8|9|aA|bB][a-fA-F0-9]{3}-[a-fA-F0-9]{12}$")
	if !reg.MatchString(rr.Body.String()) {
		t.Error("UUID was not returned")
	}
	checkResponse(t, rr, http.StatusOK)
}

func TestGetIPv4(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/ipv4", nil)
	rr := executeRequest(r, GetIPv4)
	checkResponse(t, rr, http.StatusOK)
}

func TestGetIPv6(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/ipv6", nil)
	rr := executeRequest(r, GetIPv6)
	checkResponse(t, rr, http.StatusOK)
}

func TestGetBase64(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/base64", nil)
	rr := executeRequest(r, GetBase64)
	checkResponse(t, rr, http.StatusOK)
	_, err := base64.URLEncoding.DecodeString(rr.Body.String())
	if err != nil {
		t.Errorf("Expected to decode base64 got %s", err.Error())
	}
}

func TestGETXML(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/xml", nil)
	rr := executeRequest(r, GetXML)
	checkResponse(t, rr, http.StatusOK)
	r, _ = http.NewRequest("GET", "/get/xml/5", nil)
	rr = executeVarsRequest("/get/xml/{rows}", r, GetXML)
	checkResponse(t, rr, http.StatusOK)
	r, _ = http.NewRequest("GET", "/get/xml/hello", nil)
	rr = executeVarsRequest("/get/xml/{rows}", r, GetXML)
	checkResponse(t, rr, http.StatusBadRequest)
}

func TestGetCSV(t *testing.T) {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	r, _ := http.NewRequest("GET", "/get/csv", nil)
	rr := executeRequest(r, GetCSV)
	checkResponse(t, rr, http.StatusOK)
	r, _ = http.NewRequest("GET", "/get/csv/5", nil)
	rr = executeVarsRequest("/get/csv/{rows}", r, GetCSV)
	checkResponse(t, rr, http.StatusOK)
	r, _ = http.NewRequest("GET", "/get/csv/hello", nil)
	rr = executeVarsRequest("/get/csv/{rows}", r, GetCSV)
	checkResponse(t, rr, http.StatusBadRequest)
}


auth_test.go
package main

import (
	"net/http"
	"testing"
)

func TestBadAuth(t *testing.T) {
	r, _ := http.NewRequest("GET", "/auth/basic/", nil)
	r.SetBasicAuth("user", "user")
	rr := executeRequest(r, DynamicAuth)
	checkResponse(t, rr, http.StatusUnauthorized)
}

func TestGoodAuth(t *testing.T) {
	r, _ := http.NewRequest("GET", "/auth/basic/", nil)
	r.SetBasicAuth("admin", "admin")
	rr := executeRequest(r, DynamicAuth)
	checkResponse(t, rr, http.StatusOK)
}

func TestBadDynamicAuth(t *testing.T) {
	r, _ := http.NewRequest("GET", "/auth/basic/user/bad", nil)
	r.SetBasicAuth("user", "failed")
	rr := executeVarsRequest("/auth/basic/{username}/{password}", r, DynamicAuth)
	checkResponse(t, rr, http.StatusUnauthorized)
}

func TestGoodDynamicAuth(t *testing.T) {
	r, _ := http.NewRequest("GET", "/auth/basic/user/allowed", nil)
	r.SetBasicAuth("user", "allowed")
	rr := executeVarsRequest("/auth/basic/{username}/{password}", r, DynamicAuth)
	checkResponse(t, rr, http.StatusOK)
}


main_test.go
package main

import (
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/gorilla/mux"
)

func TestHello(t *testing.T) {
	r, _ := http.NewRequest("GET", "/", nil)
	rr := executeRequest(r, Hello)
	checkResponse(t, rr, http.StatusOK)
}

func TestHeaders(t *testing.T) {
	r, _ := http.NewRequest("GET", "/headers", nil)
	r.Header.Add("hello", "world")
	rr := executeRequest(r, EchoHeaders)
	checkResponse(t, rr, http.StatusOK)
}

func TestStatusCodeGood(t *testing.T) {
	r, _ := http.NewRequest("GET", "/status", nil)
	rr := executeRequest(r, StatusCode)
	checkResponse(t, rr, http.StatusOK)

}

func TestStatusCode404(t *testing.T) {
	r, _ := http.NewRequest("GET", "/status/404", nil)
	rr := executeVarsRequest("/status/{code}", r, StatusCode)
	checkResponse(t, rr, http.StatusNotFound)
}

func TestStatusCode500(t *testing.T) {
	r, _ := http.NewRequest("GET", "/status/500", nil)
	rr := executeVarsRequest("/status/{code}", r, StatusCode)
	checkResponse(t, rr, http.StatusInternalServerError)
}

func TestStatusCodeFloat(t *testing.T) {
	r, _ := http.NewRequest("GET", "/status/5.1", nil)
	rr := executeVarsRequest("/status/{code}", r, StatusCode)
	checkResponse(t, rr, http.StatusBadRequest)
}

func TestStatusCodeString(t *testing.T) {
	r, _ := http.NewRequest("GET", "/status/fail", nil)
	rr := executeVarsRequest("/status/{code}", r, StatusCode)
	checkResponse(t, rr, http.StatusBadRequest)
}

func TestStatusCodeExtra(t *testing.T) {
	r, _ := http.NewRequest("GET", "/status/700", nil)
	rr := executeVarsRequest("/status/{code}", r, StatusCode)
	checkResponse(t, rr, 700)
}

func executeVarsRequest(path string, req *http.Request, responseFunction func(w http.ResponseWriter, r *http.Request)) *httptest.ResponseRecorder {
	rr := httptest.NewRecorder()
	router := mux.NewRouter()
	router.HandleFunc(path, responseFunction)
	router.ServeHTTP(rr, req)
	return rr
}

func executeRequest(req *http.Request, responseFunction func(w http.ResponseWriter, r *http.Request)) *httptest.ResponseRecorder {
	rr := httptest.NewRecorder()
	srv := http.HandlerFunc(responseFunction)
	srv.ServeHTTP(rr, req)
	return rr
}

func checkResponse(t *testing.T, resp *httptest.ResponseRecorder, expected int) {
	if expected != resp.Code {
		t.Errorf("Expected response code %d and got %d.\n. Response body: %s\n", expected, resp.Code, resp.Body.String())
	}
}


get.go
package main

import (
	"bytes"
	"encoding/base64"
	"fmt"
	"image/jpeg"
	"image/png"
	"io/ioutil"
	"log"

	"net/http"
	"strconv"

	common "github.com/Cyb3r-Jak3/common/go"
	"github.com/brianvoe/gofakeit/v6"
	"github.com/gorilla/mux"
)

func generaterowcount(req *http.Request) (rowCount int, rowErr error) {
	vars := mux.Vars(req)
	if vars["rows"] == "" {
		rowCount = defaultRowCount
	} else {
		rowCount, rowErr = strconv.Atoi(vars["rows"])
	}
	return rowCount, rowErr

}

// GetJSON Return random rows of JSON
func GetJSON(w http.ResponseWriter, req *http.Request) {
	rowCount, err := generaterowcount(req)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	randomdata, err := Faker.JSON(&gofakeit.JSONOptions{
		Type:     "array",
		RowCount: rowCount,
		Fields:   defaultFields,
		Indent:   true,
	})
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	common.JSONResponse(w, randomdata)
}

// downloadImage downs an image from the URL and encodes to PNG if needed
func downloadImage(url string, format string) ([]byte, error) {
	resp, err := http.Get(url) // #nosec
	if err != nil {
		return nil, err
	}
	respImage, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if err = resp.Body.Close(); err != nil {
		log.Printf("Error closing the resp body. %s\n", err)
	}
	if format == "png" {
		//Copied from https://github.com/tizz98/comix/blob/master/app/img.go
		respImage, err := jpeg.Decode(bytes.NewReader(respImage))
		if err != nil {
			return nil, err
		}

		buf := new(bytes.Buffer)
		if err := png.Encode(buf, respImage); err != nil {
			return nil, err
		}
		return buf.Bytes(), nil
	}
	return respImage, nil

}

func isvalidformat(format string) bool {
	switch format {
	case
		"jpg",
		"png",
		"url":
		return true
	}
	return false
}

// GenerateImageURL makes a valid image URL if one is not given
func GenerateImageURL(vars map[string]string) string {
	sizeOptions := [7][2]string{
		{"16", "16"},
		{"32", "32"},
		{"64", "64"},
		{"480", "360"},
		{"640", "480"},
		{"1080", "720"},
		{"1920", "1080"},
	}
	if vars["width"] == "" {
		if vars["height"] == "" {
			intChoice, _ := common.GenerateRandInt(len(sizeOptions))
			pick := sizeOptions[intChoice]
			return fmt.Sprint(baseImageURL, fmt.Sprintf("%s/%s.jpg", pick[0], pick[1]))
		}
		if isvalidformat(vars["type"]) {
			return fmt.Sprint(baseImageURL, fmt.Sprintf("%s.jpg", vars["height"]))
		}
		return fmt.Sprint(baseImageURL, fmt.Sprintf("%s/%s.jpg", vars["type"], vars["height"]))

	}
	return fmt.Sprint(baseImageURL, fmt.Sprintf("%s/%s.jpg", vars["height"], vars["width"]))

}

// GetImage downloads either a JPG, PNG or URL to an image from picsum.photos
func GetImage(w http.ResponseWriter, req *http.Request) {
	var Image []byte
	var ImageErr error
	var imageType, imageURL string
	vars := mux.Vars(req)
	if vars["type"] == "" {
		imageTypes := make([]string, 0)
		imageTypes = append(imageTypes,
			"png",
			"jpg",
			"url",
		)
		intChoice, err := common.GenerateRandInt(len(imageTypes))
		ImageErr = err
		imageType = imageTypes[intChoice]
	} else {
		imageType = vars["type"]
	}
	imageURL = GenerateImageURL(vars)
	switch imageType {
	case "png":
		Image, ImageErr = downloadImage(imageURL, imageType)
		w.Header().Set("Content-Type", "image/png")
		w.Header().Add("Content-Disposition", "attachment;filename=random.png")
	case "url":
		w.Header().Set("Content-Type", "text/plain")
		Image = []byte(imageURL)
	default:
		w.Header().Set("Content-Type", "image/jpeg")
		w.Header().Add("Content-Disposition", "attachment;filename=random.jpg")
		Image, ImageErr = downloadImage(imageURL, imageType)
	}
	if ImageErr != nil {
		http.Error(w, ImageErr.Error(), http.StatusInternalServerError)
	}
	w.Header().Set("Content-Length", fmt.Sprint(len(Image)))
	w.WriteHeader(http.StatusOK)
	if _, err := w.Write(Image); err != nil {
		log.Printf("Error writing image: %s\n", err)
	}
}

// GetUUID returns a random UUID as a string
func GetUUID(w http.ResponseWriter, req *http.Request) { common.StringResponse(w, Faker.UUID()) }

// GetIPv4 returns a random IPv4 Address
func GetIPv4(w http.ResponseWriter, _ *http.Request) { common.StringResponse(w, Faker.IPv4Address()) }

// GetIPv6 returns a random IPv6 Address
func GetIPv6(w http.ResponseWriter, _ *http.Request) { common.StringResponse(w, Faker.IPv6Address()) }

// GetBase64 return random paragraph that is base64 encoded
func GetBase64(w http.ResponseWriter, req *http.Request) {
	text := Faker.Paragraph(1, 5, 100, " ")
	encText := base64.URLEncoding.EncodeToString([]byte(text))
	common.StringResponse(w, encText)
}

//GetXML generates an XML file for a given number of rows
func GetXML(w http.ResponseWriter, req *http.Request) {
	rowCount, err := generaterowcount(req)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
	}
	randomdata, err := Faker.XML(&gofakeit.XMLOptions{
		Type:          "array",
		RootElement:   "xml",
		RecordElement: "record",
		RowCount:      rowCount,
		Indent:        true,
		Fields:        defaultFields,
	})
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	w.Header().Add("Content-Disposition", "attachment;filename=random.xml")
	common.ContentResponse(w, "text/xml", randomdata)
}

// GetCSV generates a CSV file
func GetCSV(w http.ResponseWriter, req *http.Request) {
	rowCount, err := generaterowcount(req)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
	}
	randomdata, err := Faker.CSV(&gofakeit.CSVOptions{
		Delimiter: ",",
		RowCount:  rowCount,
		Fields:    defaultFields,
	})
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	w.Header().Add("Content-Disposition", "attachment;filename=random.csv")
	common.ContentResponse(w, "text/xml", randomdata)
}


auth.go
package main

import (
	"crypto/subtle"
	"log"
	"net/http"

	common "github.com/Cyb3r-Jak3/common/go"
	"github.com/gorilla/mux"
)

const defaultAuth = "admin"

// DynamicAuth does basic authorization for given username and password
func DynamicAuth(w http.ResponseWriter, req *http.Request) {
	vars := mux.Vars(req)
	var username, password string
	if vars["username"] == "" {
		username = defaultAuth
		password = defaultAuth
	} else {
		username = vars["username"]
		password = vars["password"]
	}
	user, pass, ok := req.BasicAuth()

	if !ok || subtle.ConstantTimeCompare([]byte(user), []byte(username)) != 1 || subtle.ConstantTimeCompare([]byte(pass), []byte(password)) != 1 {
		w.Header().Set("WWW-Authenticate", `Basic realm="Simple HTTP Server"`)
		w.WriteHeader(401)
		if _, err := w.Write([]byte("Unauthorized.\n")); err != nil {
			log.Printf("Error writing authorized response: %s\n", err)
		}
		return
	}
	common.StringResponse(w, "Authenticated")
}


misc.go
package main

import (
	"fmt"
	"log"
	"net/http"
	"strconv"

	common "github.com/Cyb3r-Jak3/common/go"
	"github.com/gorilla/mux"
)

// Hello is a simple hello function
func Hello(w http.ResponseWriter, _ *http.Request) { common.StringResponse(w, "Hello") }

// EchoHeaders returns the headers of the request
func EchoHeaders(w http.ResponseWriter, req *http.Request) {

	for name, headers := range req.Header {
		for _, h := range headers {
			fmt.Fprintf(w, "%v: %v\n", name, h)
		}
	}
}

// StatusCode returns either the given code or a random one
func StatusCode(w http.ResponseWriter, req *http.Request) {
	vars := mux.Vars(req)
	if vars["code"] == "" {
		common.StringResponse(w, "Ok")
	} else {
		i, err := strconv.Atoi(vars["code"])
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
		}
		w.WriteHeader(i)
		if _, err := w.Write([]byte(fmt.Sprintf("Status Code: %d", i))); err != nil {
			log.Printf("Error writing status code response: %s\n", err)
		}
	}
}


post.go
package main

import (
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"path/filepath"

	common "github.com/Cyb3r-Jak3/common/go"
	"github.com/gorilla/mux"
)

// PostJSON echos JSON back that it was sent
func PostJSON(w http.ResponseWriter, req *http.Request) {
	req.Body = http.MaxBytesReader(w, req.Body, maxUploadSize*1024*1024)
	if req.Body == http.NoBody || req.ContentLength == 0 {
		http.Error(w, "JSON body required", http.StatusBadRequest)
		return
	}
	out, err := ioutil.ReadAll(req.Body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	common.JSONResponse(w, out)
}

// PostFormFile saves a file that upload as a form-data/multipart request
func PostFormFile(w http.ResponseWriter, req *http.Request) {
	req.Body = http.MaxBytesReader(w, req.Body, maxUploadSize*1024*1024)
	file, handler, err := req.FormFile("file")
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
	}
	f, err := os.OpenFile(filepath.Join(dirName, handler.Filename), os.O_WRONLY|os.O_CREATE, 0200)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	if _, err := io.Copy(f, file); err != nil {
		log.Printf("Error with io.Copy: %s\n", err)
		http.Error(w, "Error with copying file", http.StatusInternalServerError)
		return
	}
	common.StringResponse(w, "File uploaded")
	if err := file.Close(); err != nil {
		log.Printf("Error closing file: %s\n", err)
	}
	if err := f.Close(); err != nil {
		log.Printf("Error closing f: %s\n", err)
	}
}

//PostFile saves a file that is posted in a request body
func PostFile(w http.ResponseWriter, req *http.Request) {
	req.Body = http.MaxBytesReader(w, req.Body, maxUploadSize*1024*1024)
	vars := mux.Vars(req)
	f, err := os.OpenFile(filepath.Join(dirName, vars["name"]), os.O_WRONLY|os.O_CREATE, 0200)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	if _, err := io.Copy(f, req.Body); err != nil {
		log.Printf("Error with io.Copy: %s\n", err)
		http.Error(w, "Error with copying file", http.StatusInternalServerError)
		return
	}
	common.StringResponse(w, "File uploaded")
	if err := f.Close(); err != nil {
		log.Printf("Error closing f: %s\n", err)
	}
}


cookies.go
package main

import (
	"fmt"
	"net/http"
	"time"

	common "github.com/Cyb3r-Jak3/common/go"
	"github.com/gorilla/mux"
)

// GetCookies gets and prints all cookies from a request
func GetCookies(w http.ResponseWriter, req *http.Request) {
	for _, cookie := range req.Cookies() {
		fmt.Fprintf(w, "%v\n", cookie)
	}
}

//SetCookie sets a new cookie
func SetCookie(w http.ResponseWriter, req *http.Request) {
	vars := mux.Vars(req)
	experation := time.Now().Add(24 * time.Hour)
	cookie := &http.Cookie{
		Name:     vars["name"],
		Value:    vars["value"],
		HttpOnly: true,
		Domain:   cookieDomain,
		Expires:  experation,
		Secure:   true,
	}
	http.SetCookie(w, cookie)
}

//ClearCookies clears all the cookies from a request
func ClearCookies(w http.ResponseWriter, req *http.Request) {
	for _, cookie := range req.Cookies() {
		clearedCookie, _ := req.Cookie(cookie.Name)
		clearedCookie.MaxAge = -1
		http.SetCookie(w, clearedCookie)
	}
	common.StringResponse(w, "All Cookies Should Be Cleared")
}


main.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"path"
	"time"

	common "github.com/Cyb3r-Jak3/common/go"
	"github.com/brianvoe/gofakeit/v6"
	"github.com/gorilla/mux"
)

var host string
var port string


func hashanddelete() {
	dir, err := ioutil.ReadDir(dirName)
	if err != nil {
		log.Fatal(err)
	}
	for _, d := range dir {
		hashed, err := common.HashFile("256", path.Join([]string{dirName, d.Name()}...))
		if err != nil {
			log.Printf("Error when hasing file: %s", err)
		}
		fmt.Printf("Hash for %s: %s", d.Name(), hashed)
		log.Printf("Removing %s\n", d.Name())
		if err := os.RemoveAll(path.Join([]string{dirName, d.Name()}...)); err != nil {
			log.Printf("Error deleting file %s\n", err)
		}
	}
}

func cleardir() {
	if err := os.Mkdir(dirName, 0200); err != nil {
		log.Printf("Error creating directory %s\n", err)
	}

	for {
		hashanddelete()
		time.Sleep(cleanSeconds * time.Second)
	}
}

func init() {
	host = common.GetEnv("HOST", "")
	port = common.GetEnv("PORT", "8090")
}

func main() {
	Faker = gofakeit.NewCrypto()
	gofakeit.SetGlobalFaker(Faker)
	go cleardir()
	r := mux.NewRouter()
	r.HandleFunc("/", Hello)
	r.HandleFunc("/headers", EchoHeaders)
	r.HandleFunc("/post/json", common.AllowedMethod(PostJSON, "POST"))
	r.HandleFunc("/post/file/form", common.AllowedMethod(PostFormFile, "POST"))
	r.HandleFunc("/post/file/{name}", common.AllowedMethod(PostFile, "POST"))
	r.HandleFunc("/get/json", common.AllowedMethod(GetJSON, "GET"))
	r.HandleFunc("/get/json/{rows}", common.AllowedMethod(GetJSON, "GET"))
	r.HandleFunc("/get/image", common.AllowedMethod(GetImage, "GET"))
	r.HandleFunc("/get/image/{type}", common.AllowedMethod(GetImage, "GET"))
	r.HandleFunc("/get/image/{type}/{height}", common.AllowedMethod(GetImage, "GET"))
	r.HandleFunc("/get/image/{type}/{height}/{width}", common.AllowedMethod(GetImage, "GET"))
	r.HandleFunc("/get/uuid", common.AllowedMethod(GetUUID, "GET"))
	r.HandleFunc("/get/ipv4", common.AllowedMethod(GetIPv4, "GET"))
	r.HandleFunc("/get/ipv6", common.AllowedMethod(GetIPv6, "GET"))
	r.HandleFunc("/get/base64", common.AllowedMethod(GetBase64, "GET"))
	r.HandleFunc("/get/xml", common.AllowedMethod(GetXML, "GET"))
	r.HandleFunc("/get/xml/{rows}", common.AllowedMethod(GetXML, "GET"))
	r.HandleFunc("/get/csv", common.AllowedMethod(GetCSV, "GET"))
	r.HandleFunc("/get/csv/{rows}", common.AllowedMethod(GetCSV, "GET"))
	r.HandleFunc("/cookies/get", common.AllowedMethod(GetCookies, "GET,POST"))
	r.HandleFunc("/cookies/set/{name}/{value}", common.AllowedMethod(SetCookie, "GET,POST"))
	r.HandleFunc("/cookies/clear", common.AllowedMethod(ClearCookies, "GET,POST"))
	r.HandleFunc("/status", StatusCode)
	r.HandleFunc("/status/{code}", StatusCode)
	r.HandleFunc("/redirect", Redirect)
	r.HandleFunc("/redirect/{code}", Redirect)
	r.HandleFunc("/auth/basic/{username}/{password}", DynamicAuth)
	r.HandleFunc("/auth/basic/", DynamicAuth)
	err := http.ListenAndServe(fmt.Sprintf("%s:%s", host, port), r)
	if err != nil {
		log.Fatal(err)
	}
}
